plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java

```diff
@@ -0,0 +1,498 @@
+import org.opensearch.plugin.correlation.events.action.StoreCorrelationAction;
+import org.opensearch.plugin.correlation.events.action.StoreCorrelationRequest;
+import org.opensearch.plugin.correlation.events.action.StoreCorrelationResponse;
+import org.opensearch.plugin.correlation.rules.model.CorrelationQuery;
+import org.opensearch.plugin.correlation.rules.model.CorrelationRule;
+import org.opensearch.plugin.correlation.settings.EventsCorrelationSettings;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.SearchHits;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Transport Action for indexing correlations for a particular event
+ *
+ * @opensearch.internal
+ */
+public class TransportIndexCorrelationAction extends HandledTransportAction<IndexCorrelationRequest, IndexCorrelationResponse> {
+
+    private static final Logger log = LogManager.getLogger(TransportIndexCorrelationAction.class);
+
+    private final Client client;
+
+    private final NamedXContentRegistry xContentRegistry;
+
+    private final Settings settings;
+
+    private final ClusterService clusterService;
+
+    private volatile long correlationTimeWindow;
+
+    /**
+     * Parameterized ctor for Transport Action
+     * @param transportService TransportService
+     * @param client OS client
+     * @param xContentRegistry XContentRegistry
+     * @param settings Settings
+     * @param actionFilters ActionFilters
+     * @param clusterService ClusterService
+     */
+    @Inject
+    public TransportIndexCorrelationAction(
+        TransportService transportService,
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        Settings settings,
+        ActionFilters actionFilters,
+        ClusterService clusterService
+    ) {
+        super(IndexCorrelationAction.NAME, transportService, actionFilters, IndexCorrelationRequest::new);
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.correlationTimeWindow = EventsCorrelationSettings.CORRELATION_TIME_WINDOW.get(this.settings).getMillis();
+
+        this.clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(EventsCorrelationSettings.CORRELATION_TIME_WINDOW, it -> correlationTimeWindow = it.getMillis());
+    }
+
+    @Override
+    protected void doExecute(Task task, IndexCorrelationRequest request, ActionListener<IndexCorrelationResponse> listener) {
+        AsyncIndexCorrelationAction asyncAction = new AsyncIndexCorrelationAction(request, listener);
+        asyncAction.start();
```

Comment:
These are heavy weight transport actions chaining multiple queries as well. In which threadpool they would run. They shouldn't run on transport threadpool.

Reply:
today, all of them run in transport threadpool.